### PR TITLE
[Merged by Bors] - feat: the graded tensor product

### PR DESCRIPTION
**Pull request by eric-wieser**
This adds the graded tensor product $A \hat\otimes_R B$, imbued with a multiplication defined on homogeneous
tensors by:
$$(a \hat\otimes b) \cdot (a' \hat\otimes b') = (-1)^{\deg a' \deg b} (a \cdot a') \hat\otimes (b \cdot b')$$
where $A$ and $B$ are algebras graded by `ℕ`, `ℤ`, or `ZMod 2` (or more generally, any index that satisfies `Module ι (Additive ℤˣ)`).

We do this in two phases:

* Define bundled morphisms that perform the multiplication and braiding maps on regular tensor products of `DirectSum`s
* Define a new `GradedTensorProduct` type synonym that attaches the multiplication operator to `*` notation, and restricts itself to internally-graded algebras.

---

- [x] depends on: #7574
- [x] depends on: #7575
- [x] depends on: #7577
- [x] depends on: #7661
- [x] depends on: #7662
- [x] depends on: #7866

[Open in Gitpod](https://gitpod.io/from-referrer/)

Naming suggestions would be appreciated, I can't find a particularly good reference.

This was a project I suggested at LftCM2023 that ultimately wasn't allocated; though I think I recall that @rzeidler expressed some interest in it.

### CODE REVIEW

**Comment by eric-wieser** [comment]
!bench

**Comment by eric-wieser** [comment]
Darn, just misses the top 10 slowest files by one place

File: Mathlib/LinearAlgebra/TensorProduct/Graded.lean [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+lemma z₂pow_def (s : ℤˣ) (x : ZMod 2) : s ^ x = s ^ x.val := rfl
+
+@[simp] lemma z₂pow_zero (s : ℤˣ) : (s ^ (0 : ZMod 2) : ℤˣ) = (1 : ℤˣ) := pow_zero _
+@[simp] lemma z₂pow_one (s : ℤˣ) : (s ^ (1 : ZMod 2) : ℤˣ) = s := pow_one _
+lemma z₂pow_add (s : ℤˣ) (x y : ZMod 2) : s ^ (x + y) = s ^ x * s ^ y := by
+  simp only [z₂pow_def]
+  rw [ZMod.val_add, ←pow_eq_pow_mod, pow_add]
+  fin_cases s <;> simp
+
+end zmod2_pow
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorProduct/Graded.lean:63]: I've made [a zulip thread about this](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Powers.20of.20.60.E2.84.A4.CB.A3.60.20by.20.60ZMod.202.60/near/396291926).

File: Mathlib/Data/ZMod/IntUnitsPower.lean
Comment (eric-wieser) [inline Mathlib/Data/ZMod/IntUnitsPower.lean:1]: I've made [a zulip thread about this](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Powers.20of.20.60.E2.84.A4.CB.A3.60.20by.20.60ZMod.202.60/near/396291926).

**Review by YaelDillies — COMMENTED** [review]
This looks great and basically optimal. I have no high level picture comments, because I don't know the mathematics incredibly well, and all that I know makes me expect exactly what you did.

I have a few scattered comments, but I'm overall happy.

File: Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+variable (R) in
+/-- A Type synonym for `A ⊗[R] B`, but with multiplication as `TensorProduct.gradedMul`.
+
+This has notation `𝒜 ᵍ⊗[R] ℬ`. -/
+@[nolint unusedArguments]
+def GradedTensorProduct
+    (𝒜 : ι → Submodule R A) (ℬ : ι → Submodule R B)
+    [GradedAlgebra 𝒜] [GradedAlgebra ℬ] :
+    Type _ :=
+  A ⊗[R] B
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean:73]: I've left this as a potential TODO in the module comment. Irreducibility would make a lot of things more painful here I think, and I'd rather not pay that cost until there is more than one downstream file using it.

File: Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+$$(a \otimes b) \cdot (a' \otimes b') = (-1)^{\deg a' \deg b} (a \cdot a') \otimes (b \cdot b')$$
+
+where $A$ and $B$ are algebras graded by `ℕ`, `ℤ`, or `ι` (or more generally, any index
+that satisfies `Module ι (Additive ℤˣ)`).
+
+## Main results
+
+* `GradedTensorProduct R 𝒜 ℬ`: for families of submodules of `A` and `B` that form a graded algebra,
+  this is a type alias for `A ᵍ⊗[R] B` with the appropriate multiplication.
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean]: Done.

File: Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+The results for internally-graded algebras (via `GradedAlgebra`) are elsewhere, as is the type
+`GradedTensorProduct`.
+
+## Main results
+
+* `TensorProduct.gradedComm`: the symmetric braiding operator on the tensor product of
+  externally-graded rings.
+* `TensorProduct.gradedMul`: the previously describe multiplication on externally-graded rings, as a
+  bilinear map.
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean:31]:
```suggestion
* `TensorProduct.gradedMul`: the previously described multiplication on externally-graded rings, as
  a bilinear map.
```

File: Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+open DirectSum (lof)
+
+variable (R)
+
+section gradedComm
+
+local notation "𝒜ℬ" => (fun i : ι × ι => 𝒜 (Prod.fst i) ⊗[R] ℬ (Prod.snd i))
+local notation "ℬ𝒜" => (fun i : ι × ι => ℬ (Prod.fst i) ⊗[R] 𝒜 (Prod.snd i))
+
+
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean:75]: Not sure this was intentional
```suggestion

```

File: Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+theorem gradedComm_algebraMap (r : R) :
+    gradedComm R 𝒜 ℬ (algebraMap R _ r) = algebraMap R _ r :=
+  (gradedComm_algebraMap_tmul R 𝒜 ℬ r 1).trans (Algebra.TensorProduct.algebraMap_apply' r).symm
+
+end gradedComm
+
+/-- The multiplication operation for tensor products of externally `ι`-graded algebras. -/
+noncomputable irreducible_def gradedMul :
+    letI AB := (DirectSum _ 𝒜) ⊗[R] (DirectSum _ ℬ)
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean:175]: Are the brackets necessary?
```suggestion
    letI AB := DirectSum _ 𝒜 ⊗[R] DirectSum _ ℬ
```

File: Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+  suffices LinearMap.llcomp R _ _ _ mA ∘ₗ mA =
+      (LinearMap.llcomp R _ _ _ LinearMap.lflip <| LinearMap.llcomp R _ _ _ mA.flip ∘ₗ mA).flip by
+    exact FunLike.congr_fun (FunLike.congr_fun (FunLike.congr_fun this x) y) z
+  ext ixa xa ixb xb iya ya iyb yb iza za izb zb
+  dsimp
+  simp_rw [tmul_of_gradedMul_of_tmul, Units.smul_def, zsmul_eq_smul_cast R,
+    LinearMap.map_smul₂, LinearMap.map_smul, DirectSum.lof_eq_of, DirectSum.of_mul_of,
+    ← DirectSum.lof_eq_of R, tmul_of_gradedMul_of_tmul, DirectSum.lof_eq_of, ← DirectSum.of_mul_of,
+    ← DirectSum.lof_eq_of R, mul_assoc]
+  save
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean:245]: I assume this was for development only and that we shouldn't be committing `save` in final code?
```suggestion
```

File: Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+
+@[inherit_doc]
+notation:100 x " ᵍ⊗ₜ" y:100 => tmul _ x y
+
+@[inherit_doc]
+notation:100 x " ᵍ⊗ₜ[" R "] " y:100 => tmul R x y
+
+variable (R) in
+/-- An auxiliary construction to move between the graded tensor product of internally-graded objects
+and the tensor product of direct sums.-/
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean:124]:
```suggestion
and the tensor product of direct sums. -/
```

**Review by YaelDillies — COMMENTED** [review]
This is peak Eric code. I can't find anything to point out but typos!

**Review by YaelDillies — APPROVED** [review]
LGTM

maintainer merge

**Comment by robertylewis** [comment]
bors merge

**Pull request closed without merging**